### ISSUE TITLE
Cannot find rocm library hip_hcc

### ISSUE THREAD

**Issue by acxz** [rocm-arch/tensorflow-rocm#5]
```
Repository rule rocm_configure defined at:
  /home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl:861:33: in <toplevel>
ERROR: An error occurred during the fetch of repository 'local_config_rocm':
   Traceback (most recent call last):
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 840
		_create_local_rocm_repository(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 647, in _create_local_rocm_repository
		_find_libs(repository_ctx, <2 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 449, in _find_libs
		_select_rocm_lib_paths(<3 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 418, in _select_rocm_lib_paths
		auto_configure_fail(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 162, in auto_configure_fail
		fail(<1 more arguments>)

ROCm Configuration Error: Cannot find rocm library hip_hcc
INFO: Repository com_google_protobuf instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule tf_http_archive defined at:
  /home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/repo.bzl:134:34: in <toplevel>
ERROR: Skipping '//tensorflow:install_headers': no such package '@local_config_rocm//rocm': Traceback (most recent call last):
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 840
		_create_local_rocm_repository(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 647, in _create_local_rocm_repository
		_find_libs(repository_ctx, <2 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 449, in _find_libs
		_select_rocm_lib_paths(<3 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 418, in _select_rocm_lib_paths
		auto_configure_fail(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 162, in auto_configure_fail
		fail(<1 more arguments>)

ROCm Configuration Error: Cannot find rocm library hip_hcc
ERROR: no such package '@local_config_rocm//rocm': Traceback (most recent call last):
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 840
		_create_local_rocm_repository(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 647, in _create_local_rocm_repository
		_find_libs(repository_ctx, <2 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 449, in _find_libs
		_select_rocm_lib_paths(<3 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 418, in _select_rocm_lib_paths
		auto_configure_fail(<1 more arguments>)
	File "/home/acxz/vcs/git/github/rocm-arch/tensorflow-rocm/src/tensorflow-2.2.0-rocm/third_party/gpus/rocm_configure.bzl", line 162, in auto_configure_fail
		fail(<1 more arguments>)

ROCm Configuration Error: Cannot find rocm library hip_hcc
INFO: Elapsed time: 5.658s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
    currently loading: tensorflow ... (2 packages)
==> ERROR: A failure occurred in build().
    Aborting...
```

**Comment by jtiemer** [comment]
I came across this too. Let me know if I can support you with additional info. I am very interested in having this package build properly. The same error is raised when trying to build pytorch-rocm by the way.

**Comment by acxz** [comment]
Yeah I've been stuck on this one for a while, would appreciate any help I can get, prob most likely to with the environment variables. I'm focusing on getting pytorch-rocm for now.
If you are experiencing this in the pytorch-rocm pkgbuild, can you make a separate issue in that repo so that we can track it?

**Comment by jtiemer** [comment]
Will do. Would you consider building through an AUR-helper (pikaur in my case) legit or should I build in a totally clean root with no helper? In my experience pikaur works fine wrt cleanliness etc.
What is hip_hcc anyway? According to the little research I did, it may be the deprecated "hcc" part of the AMD ROCm stack?

**Comment by acxz** [comment]
For development you should build in a clean root, but I only build in a clean chroot if I can get it building without a clean chroot first lol. 
I rec `yay` instead of `pikaur` tho, just cause I have had bugs with `pikaur` in the past and it's better if we use the same tools as much as possible. i.e. if you have a problem with regular `makepkg` I'll help debug, if you use what I use `yay` I'll help debug, but if you have a problem on `pikaur` and I can't reproduce it, then gg.

Yeah tf might still be using the older hcc idk know tho, haven't really looked into this problem.

**Comment by Imajie** [comment]
It looks like an issue with the default path that it's looking for hip_hcc to be on the system. The other PKGBUILDs install it to /opt/rocm/hip, but the bazel build script here is looking at just /opt/rocm.

If I add a sed to the PKGBUILD in `prepare()` before it copies the src folders for the different flavors to add the appropriate suffix to the path it seems to be building fine now:
```
sed -i 's/\("hip_hcc", rocm_config.rocm_toolkit_path\)/\1 + "\/hip"/g' tensorflow-${_pkgver}/third_party/gpus/rocm_configure.bzl
```

**Comment by acxz** [comment]
o.O nice find! Can you make a patch instead of a `sed` command in the PKGBUILD, and make a PR so that others can test your changes by just pulling your PR?

Also do you think this is an upstream issue with tensorflow or should the issue be tackled in rocm-arch? Just from a glance it seems it should be an upstream tensorflow issue right?

**Comment by Imajie** [comment]
Update: the build goes for a while but fails with a few other issues later on: it has an issue finding an miopen include file and fails to run a hip command, seems like a path issue again.

I can make a patch to include in the PKGBUILD and put up a PR for that, but the build still seems to fail later on

**Comment by acxz** [comment]
Any progress is good progress! Although if this is an upstream issue, either rocm-arch or tensorflow, we do need to make an issue/PR at the correct repository before I merge the PR.

I went ahead and made an upstream issue so that we can merge our patch in
https://github.com/tensorflow/tensorflow/issues/42081

upstream patch merged waiting on new release to close this issue: https://github.com/tensorflow/tensorflow/pull/42095

**Comment by jtiemer** [comment]
Checking GPU availability fails with:

"Could not load dynamic library `'libhip_hcc.so'`":

```
$ python
Python 3.8.5 (default, Jul 27 2020, 08:42:51) 
[GCC 10.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.test.gpu_device_name()
2020-08-20 00:00:11.956125: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
2020-08-20 00:00:11.961707: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 3601440000 Hz
2020-08-20 00:00:11.962015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55dd52949120 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-20 00:00:11.962031: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-08-20 00:00:11.962962: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libhip_hcc.so'; dlerror: libhip_hcc.so: cannot open shared object file: No such file or directory
2020-08-20 00:00:11.962972: F tensorflow/stream_executor/lib/statusor.cc:34] Attempting to fetch value instead of handling error Failed precondition: Could not load dynamic library 'libhip_hcc.so'; dlerror: libhip_hcc.so: cannot open shared object file: No such file or directory
Canceled // Abgebrochen (wrote memory dump // Speicherabzug geschrieben)
```
So … the `libhip_hcc`-Problem might just have resurfaced.

**Comment by jtiemer** [comment]
Related 1 https://github.com/RadeonOpenCompute/ROCm/issues/1163
Related 2 https://github.com/tensorflow/tensorflow/issues/40751

Solution from 2 works for me. That is: TF GPU Test is successful:

```
import tensorflow as tf
with tf.device('/gpu:0'):
    a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
    b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
    c = tf.matmul(a, b)
```
<details>
<summary>The output is extremely verbose though. No idea why.</summary>

```
2020-08-20 00:23:46.794816: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libhip_hcc.so
2020-08-20 00:23:47.230378: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1579] Found device 0 with properties: 
pciBusID: 0000:28:00.0 name: Vega 10 XL/XT [Radeon RX Vega 56/64]     ROCm AMD GPU ISA: gfx900
coreClock: 1.59GHz coreCount: 56 deviceMemorySize: 7,98GiB deviceMemoryBandwidth: -1B/s
2020-08-20 00:23:47.375092: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocblas.so
2020-08-20 00:23:47.375988: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libMIOpen.so
2020-08-20 00:23:47.379274: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocfft.so
2020-08-20 00:23:47.380102: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocrand.so
2020-08-20 00:23:47.380206: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1703] Adding visible gpu devices: 0
2020-08-20 00:23:47.380585: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
2020-08-20 00:23:47.384817: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 3601000000 Hz
2020-08-20 00:23:47.385050: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55b44ec5b910 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-20 00:23:47.385060: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-08-20 00:23:47.385987: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55b44ed58220 initialized for platform ROCM (this does not guarantee that XLA will be used). Devices:
2020-08-20 00:23:47.385993: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Vega 10 XL/XT [Radeon RX Vega 56/64], AMDGPU ISA version: gfx900
2020-08-20 00:23:47.386088: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1579] Found device 0 with properties: 
pciBusID: 0000:28:00.0 name: Vega 10 XL/XT [Radeon RX Vega 56/64]     ROCm AMD GPU ISA: gfx900
coreClock: 1.59GHz coreCount: 56 deviceMemorySize: 7,98GiB deviceMemoryBandwidth: -1B/s
2020-08-20 00:23:47.386122: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocblas.so
2020-08-20 00:23:47.386135: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libMIOpen.so
2020-08-20 00:23:47.386146: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocfft.so
2020-08-20 00:23:47.386158: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocrand.so
2020-08-20 00:23:47.386218: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1703] Adding visible gpu devices: 0
2020-08-20 00:23:47.386233: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-20 00:23:47.386235: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 
2020-08-20 00:23:47.386238: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N 
2020-08-20 00:23:47.386315: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1247] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7384 MB memory) -> physical GPU (device: 0, name: Vega 10 XL/XT [Radeon RX Vega 56/64], pci bus id: 0000:28:00.0)
2020-08-20 00:23:47.745530: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library librocblas.so
```

</details>

```
In[2]:    c
Out[2]: 
<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[22., 28.],
       [49., 64.]], dtype=float32)>
```

**Comment by acxz** [comment]
```
echo 'export LD_LIBRARY_PATH=/opt/rocm/lib:$LD_LIBRARY_PATH' >> $HOME/.bashrc 
source $HOME/.bashrc
sudo reboot
```

Hmm I feel like this should go in the pkgbuild of hip-rocclr as the solution

**Comment by jtiemer** [comment]
Absolutely. Sorry for not copypasting the solution rightaway. It was like 2am here…

**Comment by kk-1** [comment]
Hello Folks,
I had the same problem.
I saw that the library libhip_hcc is moved to (as symlink!!!):
/usr/lib/aomp/lib

Here 2 ways:
1-) You can create symlink to "libamdhip64.so" in the PATH manually or
2-) You can install the aomp and set the LD_LIBRARY_PATH in your .profile like that (got some extras but just get the idea!):


`export LD_LIBRARY_PATH="/usr/lib/aomp/lib:/usr/lib/aomp/lib64:/opt/rocm/lib:/opt/rocm/lib64:/opt/rocm/opencl/lib/x86_64:$LD_LIBRARY_PATH"`

------------------------------------------

BTW I got:
openSUSE Tumbleweed: rpm based system
rocm 3.7.0
aomp 11.8

Cheers!!!

**Comment by acxz** [comment]
hmm  interesting thanks for pointing that out @kk-1 
For right now, I'll put a hold on this till we get the rocm 3.7 stack working on Arch
Edit: yep can confirm
Made an upstream issue about this, looks like TF is not at ROCm 3.7.0 yet

**Comment by acxz** [comment]
The build issue should be patched with 73b5b5c

With rocm 3.7 hip_hcc is removed and thus the runtime error you faced @jtiemer might not exist anymore. Can you see if you can reproduce the error with the latest rocm and tensorflow-rocm packages on Arch, @jtiemer? Once I get confirmation from you I'll remove the runtime label.

**Comment by jtiemer** [comment]
Yep, can confirm it works now.
I will still have to check whether this is due to the hack from earlier, but it doesn't complain about `libhip_hcc.so` anymore. So … I'm optimistic about it.

**Comment by acxz** [comment]
Ah plz let me know when you remove the hack from before as well, but I'm glad libhip_hcc is not a problem anymore. Thx for checking :)

**Comment by acxz** [comment]
closed with 43d4b3d
